repo/blob/sftp/sftp_storage.go

```diff
@@ -0,0 +1,338 @@
+						BlobID:    fullID,
+						Length:    e.Size(),
+						Timestamp: e.ModTime(),
+					}); err != nil {
+						return err
+					}
+				}
+			}
+		}
+
+		return nil
+	}
+
+	return walkDir(s.Path, "")
+}
+
+func (s *sftpStorage) PutBlob(ctx context.Context, blobID blob.ID, data []byte) error {
+	_, filePath := s.getDirPathAndFilePath(blobID)
+	tmpPath := fmt.Sprintf("%v-%v", filePath, rand.Int63())
+
+	f, err := s.createTempFileAndDir(tmpPath)
+	if err != nil {
+		fmt.Printf("putblob-createtempfileanddir(%s)\n", err)
+		return errors.Wrap(err, "cannot create temporary file")
+	}
+
+	if _, err = f.Write(data); err != nil {
+		fmt.Printf("putblob-write(%s)\n", err)
+		return errors.Wrap(err, "can't write temporary file")
+	}
+
+	if err = f.Close(); err != nil {
+		return errors.Wrap(err, "can't close temporary file")
+	}
+
+	err = s.cli.Rename(tmpPath, filePath)
+	if err != nil {
+		if removeErr := s.cli.Remove(tmpPath); removeErr != nil {
+			fmt.Printf("warning: can't remove temp file: %v", removeErr)
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (s *sftpStorage) DeleteBlob(ctx context.Context, blobID blob.ID) error {
+	_, filePath := s.getDirPathAndFilePath(blobID)
+	err := s.cli.Remove(filePath)
+	if err == nil || os.IsNotExist(err) {
+		return nil
+	}
+
+	return err
+}
+
+func (s *sftpStorage) getShardDirectory(blobID blob.ID) (string, blob.ID) {
+	shardPath := s.Path
+	if len(blobID) < 20 {
+		return shardPath, blobID
+	}
+	for _, size := range s.shards() {
+		shardPath = filepath.Join(shardPath, string(blobID[0:size]))
+		blobID = blobID[size:]
+	}
+
+	return shardPath, blobID
+}
+
+func (s *sftpStorage) getDirPathAndFilePath(blobID blob.ID) (dirPath, filePath string) {
+	shardPath, blobID := s.getShardDirectory(blobID)
+	result := filepath.Join(shardPath, makeFileName(blobID))
+	return shardPath, result
+}
+
+func (s *sftpStorage) createTempFileAndDir(tempFile string) (*psftp.File, error) {
+	flags := os.O_CREATE | os.O_WRONLY | os.O_EXCL
+	f, err := s.cli.OpenFile(tempFile, flags)
+	if os.IsNotExist(err) {
+		if err = s.cli.MkdirAll(filepath.Dir(tempFile)); err != nil {
+			return nil, errors.Wrap(err, "cannot create directory")
+		}
+		return s.cli.OpenFile(tempFile, flags)
+	}
+
+	return f, err
+}
+
+func (s *sftpStorage) ConnectionInfo() blob.ConnectionInfo {
+	return blob.ConnectionInfo{
+		Type:   sftpStorageType,
+		Config: &s.Options,
+	}
+}
+
+func (s *sftpStorage) Close(ctx context.Context) error {
+	s.cli.Close()
+	s.conn.Close()
+
+	return nil
+}
+
+// getHostKey parses OpenSSH known_hosts file for a public key that matches the host
+func getHostKey(host, knownHosts string) (ssh.PublicKey, error) {
+	file, err := os.Open(knownHosts)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	var hostKey ssh.PublicKey
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		fields := strings.Split(scanner.Text(), " ")
+
+		if len(fields) != 3 {
+			continue
+		}
+
+		if strings.Contains(fields[0], host) {
```

Comment:
is that substring match intentional? I think it will match all hosts if "" is provided.

Reply:
host is a required argument, can't be empty.
The check is there to extract the public key for the specific host, if it's there.
If it's not there, that means the user doesn't have a working public/private key connection to the ssh server.

Reply:
wouldn't this have side effects if hou had hosts "foo", "foo-a" and "foo-b". Now strings.Contains("foo-a", "foo") is true, but that's not indended, I think.

Reply:
I implement more robust logic, leveraging ssh.ParseKnownHosts go function.

This requires the hostname to be exactly equal to the host entry found in known_hosts.

There maybe some edge cases that this code isn't taking into account (hashed entries, wildcards, etc), but that should be on the users to solve, since they have full control over the --host, --keyfile params, as well as the known_hosts file.